src/main/java/marquez/common/models/DatasourceUrn.java

```diff
@@ -1,24 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package marquez.common.models;
 
-import java.util.regex.Pattern;
+import static marquez.common.Preconditions.checkNotBlank;
+import static marquez.common.models.UrnPattern.URN_DELIM;
+import static marquez.common.models.UrnPattern.URN_PREFIX;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.StringJoiner;
 import lombok.EqualsAndHashCode;
+import lombok.Getter;
 import lombok.NonNull;
+import lombok.ToString;
+
+@EqualsAndHashCode
+@ToString
+public final class DatasourceUrn {
+  private static final int URN_PARTS = 2;
+  private static final UrnPattern URN_PATTERN = UrnPattern.of(UrnType.DATASOURCE, URN_PARTS);
 
-@EqualsAndHashCode(callSuper = false)
-public final class DatasourceUrn extends Urn {
-  private static final int NUM_COMPONENTS = 2;
-  private static final String URN_TYPE = "datasource";
-  private static final Pattern REGEX = buildPattern(URN_TYPE, NUM_COMPONENTS);
+  @Getter private final String value;
 
-  public DatasourceUrn(@NonNull String value) {
-    super(value, REGEX);
+  private DatasourceUrn(@NonNull final String value) {
+    URN_PATTERN.throwIfNoMatch(checkNotBlank(value));
+    this.value = value;
   }
 
-  public static DatasourceUrn from(@NonNull String type, @NonNull String name) {
-    final String value = fromComponents(URN_TYPE, type, name);
+  public static DatasourceUrn from(
```

Comment:
why not build the string within `UrnPattern` as well since `UrnPattern` is enforcing a consistent format with `throwIfNotMatch`?

Reply:
Depends on the definition of `UrnPattern`. I view it more as: _representing and enforcing a pattern_ vs building the URN string. That is, with patterns, we ask does string `S` match pattern `P`. Combining functionality would clean the code quite a bit. Maybe `UrnPattern.buildOrthrowIfNotMatch()` would make sense?

Reply:
Really appreciate the comment, @sshah-wework! It inspired a much cleaner approach that both enforces and builds a URN string in the base Urn class. That is, with the introduction of the method `pattern()`,  we can enforce Urn validation in a single place (on Urn instantiation!)

```java
protected Urn(@NonNull final String value) {
  pattern().throwIfNoMatch(checkNotBlank(value));
  this.value = value;
}
```